sw/device/lib/dif/dif_spi_host.c

```diff
@@ -0,0 +1,383 @@
+  uint32_t reg = 0;
+  reg = bitfield_field32_write(reg, SPI_HOST_COMMAND_LEN_FIELD, length - 1);
+  reg = bitfield_field32_write(reg, SPI_HOST_COMMAND_SPEED_FIELD, speed);
+  reg =
+      bitfield_field32_write(reg, SPI_HOST_COMMAND_DIRECTION_FIELD, direction);
+  reg = bitfield_bit32_write(reg, SPI_HOST_COMMAND_CSAAT_BIT, !last_segment);
+  mmio_region_write32(spi_host->base_addr, SPI_HOST_COMMAND_REG_OFFSET, reg);
+}
+
+static void issue_opcode(const dif_spi_host_t *spi_host,
+                         dif_spi_host_segment_t *segment, bool last_segment) {
+  wait_tx_fifo(spi_host);
+  mmio_region_write8(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET,
+                     segment->opcode);
+  write_command_reg(spi_host, 1, kDifSpiHostWidthStandard,
+                    kDifSpiHostDirectionTx, last_segment);
+}
+
+static void issue_address(const dif_spi_host_t *spi_host,
+                          dif_spi_host_segment_t *segment, bool last_segment) {
+  wait_tx_fifo(spi_host);
+  // The address appears on the wire in big-endian order.
+  uint32_t address = bitfield_byteswap32(segment->address.address);
+  uint32_t length;
+  if (segment->address.mode == kDifSpiHostAddrMode4b) {
+    length = 4;
+    mmio_region_write32(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET,
+                        address);
+  } else {
+    length = 3;
+    address >>= 8;
+    mmio_region_write32(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET,
+                        address);
+  }
+  write_command_reg(spi_host, length, segment->address.width,
+                    kDifSpiHostDirectionTx, last_segment);
+}
+
+static void issue_dummy(const dif_spi_host_t *spi_host,
+                        dif_spi_host_segment_t *segment, bool last_segment) {
+  write_command_reg(spi_host, segment->dummy.length, segment->dummy.width,
+                    kDifSpiHostDirectionDummy, last_segment);
+}
+
+static void issue_data_phase(const dif_spi_host_t *spi_host,
+                             dif_spi_host_segment_t *segment,
+                             bool last_segment) {
+  size_t length;
+  dif_spi_host_width_t width;
+  dif_spi_host_direction_t direction;
+
+  switch (segment->type) {
+    case kDifSpiHostSegmentTypeTx:
+      width = segment->tx.width;
+      length = segment->tx.length;
+      direction = kDifSpiHostDirectionTx;
+      spi_host_fifo_write_alias(spi_host, segment->tx.buf, segment->tx.length);
+      break;
+    case kDifSpiHostSegmentTypeBidirectional:
+      width = segment->bidir.width;
+      length = segment->bidir.length;
+      direction = kDifSpiHostDirectionBidirectional;
+      spi_host_fifo_write_alias(spi_host, segment->bidir.txbuf,
+                                segment->bidir.length);
+      break;
+    case kDifSpiHostSegmentTypeRx:
+      width = segment->rx.width;
+      length = segment->rx.length;
+      direction = kDifSpiHostDirectionRx;
+      break;
+    default:
+      // Programming error.  We should never get here.
```

Comment:
Probably worth qualifying this as it being an error within this file, rather than API misuse.

Reply:
Done.